router/state_broadcast.go

```diff
@@ -0,0 +1,169 @@
+// from the actor that owns them, in order to prevent data races.
+
+// _maintainBroadcasts sends out wakeup broadcasts to let local nodes know
+// of our presence in the network.
+func (s *state) _maintainBroadcasts() {
+	select {
+	case <-s.r.context.Done():
+		return
+	default:
+		defer s._sendBroadcastIn(wakeupBroadcastInterval)
+	}
+
+	// Clean up any broadcasts that are older than the expiry period.
+	for k, v := range s._seenBroadcasts {
+		if !v.valid() {
+			s.r.log.Println("delete old broadcast")
+			delete(s._seenBroadcasts, k)
+		}
+	}
+
+	s._sendWakeupBroadcasts()
+}
+
+func (s *state) _createBroadcastFrame() (*types.Frame, error) {
+	// Construct the broadcast packet.
+	b := frameBufferPool.Get().(*[types.MaxFrameSize]byte)
+	defer frameBufferPool.Put(b)
+	broadcast := types.WakeupBroadcast{
+		Sequence: types.Varu64(time.Now().UnixMilli()),
+		Root:     s._rootAnnouncement().Root,
+	}
+	if s.r.secure {
+		protected, err := broadcast.ProtectedPayload()
+		if err != nil {
+			return nil, err
+		}
+		copy(
+			broadcast.Signature[:],
+			ed25519.Sign(s.r.private[:], protected),
+		)
+	}
+	n, err := broadcast.MarshalBinary(b[:])
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the frame.
+	send := getFrame()
+	send.Type = types.TypeWakeupBroadcast
+	send.SourceKey = s.r.public
+	send.HopLimit = types.NetworkHorizonDistance
+	send.Payload = append(send.Payload[:0], b[:n]...)
+
+	return send, nil
+}
+
+func (s *state) _sendWakeupBroadcasts() {
+	broadcast, err := s._createBroadcastFrame()
+	if err != nil {
+		s.r.log.Println("Failed creating broadcast frame:", err)
+	}
+
+	s._flood(s.r.local, broadcast, ClassicFlood)
+}
+
+func (s *state) _handleBroadcast(p *peer, f *types.Frame) error {
+	// Unmarshall the broadcast
+	var broadcast types.WakeupBroadcast
+	if _, err := broadcast.UnmarshalBinary(f.Payload); err != nil {
+		return fmt.Errorf("broadcast unmarshal failed: %w", err)
+	}
+
+	if s.r.secure {
+		// Check that the broadcast message was protected by the node that claims
+		// to have sent it. Silently drop it if there's a signature problem.
+		protected, err := broadcast.ProtectedPayload()
+		if err != nil {
+			return fmt.Errorf("broadcast payload invalid: %w", err)
+		}
+		if !ed25519.Verify(
+			f.SourceKey[:],
+			protected,
+			broadcast.Signature[:],
+		) {
+			return fmt.Errorf("broadcast payload signature invalid")
+		}
+	}
+
+	// Check that the root key in the update matches our current root.
+	// If they don't match, silently drop the broadcast.
+	root := s._rootAnnouncement()
+	if root.RootPublicKey.CompareTo(broadcast.RootPublicKey) != 0 {
+		return nil
+	}
+
+	// Check the sequence number in the broadcast.
+	// If we have seen a higher sequence number before then there is no need
+	// to continue forwarding it.
+	if existing, ok := s._seenBroadcasts[f.SourceKey]; ok {
+		// TODO : add timestamp filtering to prevent flooding DoS attacks
```

Comment:
On timestamp filtering, agreed, we probably want to make sure that we haven't received a broadcast from this peer too recently. Since we're already storing the timestamp we saw the last broadcast, it should be easy enough to do with `time.Since`.